torch/distributed/_spmd/gm_transformation.py

```diff
@@ -0,0 +1,131 @@
+import operator
+from typing import Any, Callable, Dict, List, Optional
+
+import torch
+import torch.nn as nn
+
+from functorch import make_fx
+from torch import fx
+from torch._inductor.compile_fx import compile_fx_inner
+from torch._inductor.decomposition import select_decomp_table
+from torch.distributed._spmd.graph_utils import OP
+from torch.distributed._spmd.iter_graph_module import IterGraphModule
+from torch.utils._pytree import tree_flatten, tree_map_only
+
+
+class InductorWrapper(nn.Module):
+    def __init__(self, gm: fx.GraphModule, enable_cudagraphs: bool) -> None:
+        super().__init__()
+        self._gm = gm
+        self._compiled: Optional[nn.Module] = None
+        self._enable_cudagraphs = enable_cudagraphs
+
+    def forward(self, *args: Any) -> Any:
+        if self._compiled is None:
+            gm = make_fx(self._gm, decomposition_table=select_decomp_table())(*args)
+            self._compiled = compile_fx_inner(
+                gm,
+                list(args),
+                cudagraphs=self._enable_cudagraphs,
+            )
+        list_args, _ = tree_flatten(args)
+        return self._compiled(list_args)
+
+
+def lower_to_inductor(
+    gm: torch.fx.GraphModule, enable_cudagraphs: bool
+) -> torch.fx.GraphModule:
+    orig_placeholders = []
+    orig_output_args: List[Any] = []
+    output = next(iter(gm.graph.nodes))
+    move_nodes = []
+
+    for node in gm.graph.nodes:
+        if node.op == OP.OUTPUT:
+            output = node
+            orig_output_args, _ = tree_flatten(node.args)
+        elif node.op == OP.PLACEHOLDER:
+            orig_placeholders.append(node)
+        else:
+            move_nodes.append(node)
+    orig_placeholders_set = set(orig_placeholders)
+
+    subgraph: torch.fx.Graph = torch.fx.Graph()
+    node_mapping: Dict[torch.fx.Node, torch.fx.Node] = {}
+    attrs = {}
+    for i, p in enumerate(orig_placeholders):
+        node_mapping[p] = subgraph.placeholder(name=f"placeholder_{i}")
+
+    for node in move_nodes:
+        if node.op == OP.GET_ATTR:
+            attrs[node.target] = getattr(gm, node.target)
+        args, kwargs = tree_map_only(
+            fx.Node, lambda n: node_mapping[n], (node.args, node.kwargs)
+        )
+        node_mapping[node] = subgraph.create_node(
+            op=node.op, target=node.target, args=args, kwargs=kwargs
+        )
+        node_mapping[node].meta = node.meta
+
+    placeholders: List[torch.fx.Node] = []
+    for placeholder in orig_placeholders:
+        new_placeholder = node_mapping[placeholder]
+        if len(new_placeholder.users) == 0:
+            # Remove unused placeholders from the subgraph. This is required as
+            # the `train_step()` has module and optimizer as the inputs which
```

Comment:
(no need to change this PR) in this case, should we do this right after getting the local train_step graph instead of dtensor expansion and other graph transformations?

Reply:
Users will still pass model and optimizer, right? So do you mean that we directly wrap the FX graph right after expansion?

Reply:
> No, I was referring to before expansion. 

Basically, trace local graph -> remove unused placeholders -> DTensor expansion

Because expansion is graph-traversal based solution for now, and does not need to pass in models and optimziers. 

But @wanchaol's new SPMD expansion might potentially change that, suppose we do it through our dedicated tracer.

Reply:
Still a little confused about this. A user has to call `train_step(model, optim, batch)` to trace the single step and expand the graph. The user will expect that the subsequent calls are still `train_step(model, optim, batch)` which means `model` and `optim` are the placeholders for the `fx.Graph`. We have to create a subgraph in order to drop `model` and `optim`. That subgraph can then be lowered to the Inductor. We can do this wrapping during or right after expansion which makes sense to me. But I want to confirm my understanding that the wrapping, creating a subgraph, is still needed?